components/raftstore/src/store/fsm/store.rs

```diff
@@ -1596,45 +1597,45 @@ impl<'a, EK: KvEngine, ER: RaftEngine, T: Transport, C: PdClient>
             return Ok(());
         }
         let check_msg_status = self.check_msg(&msg)?;
-        let is_first_request_vote = match check_msg_status {
+        let is_first_request = match check_msg_status {
             CheckMsgStatus::DropMsg => return Ok(()),
-            CheckMsgStatus::FirstRequestVote => true,
+            CheckMsgStatus::FirstRequest => true,
             CheckMsgStatus::NewPeer | CheckMsgStatus::NewPeerFirst => {
-                if !self.maybe_create_peer(
+                if self.maybe_create_peer(
                     region_id,
                     &msg,
                     check_msg_status == CheckMsgStatus::NewPeerFirst,
                 )? {
-                    if !util::is_first_vote_msg(msg.get_message()) {
-                        // Can not create peer from the message and it's not the
-                        // first request vote message.
-                        return Ok(());
-                    }
-                    true
-                } else {
-                    false
+                    // Peer created, send the message again
+                    let _ = self.ctx.router.send(region_id, PeerMsg::RaftMessage(msg));
+                    return Ok(());
                 }
+                // Can't create peer, see if we should keep this message
+                util::is_first_message(msg.get_message())
             }
         };
-        if is_first_request_vote {
-            // To void losing request vote messages, either put it to
-            // pending_votes or force send.
+        if is_first_request {
+            // To void losing messages, either put it to pending_msg or force send.
             let mut store_meta = self.ctx.store_meta.lock().unwrap();
             if !store_meta.regions.contains_key(&region_id) {
-                store_meta.pending_votes.push(msg);
-                return Ok(());
-            }
-            if let Err(e) = self
-                .ctx
-                .router
-                .force_send(region_id, PeerMsg::RaftMessage(msg))
-            {
-                warn!("handle first request vote failed"; "region_id" => region_id, "error" => ?e);
+                // Save one pending message for a peer is enough, remove
+                // the previous pending message of this peer
+                store_meta
+                    .pending_msgs
+                    .swap_remove_front(|m| m.get_to_peer() == msg.get_to_peer());
```

Comment:
I suggest using an independent vector to save the first append msg when `to_peer` is a learner.